### DISCUSSION TITLE
How to accomplish per-pixel depth/z buffering?

### DISCUSSION THREAD

**Discussion by JoshuaWise** [pixijs/pixijs#8380] [Q&A]
Hello!

I'm new to PixiJS, and I've been reading the guides and documentation, and it seems PixiJS supports depth sorting on a per-object basis (via `Container.sortableChildren`). However, this isn't a good solution for many types of games, where objects can intersect each other, or wrap around each other (often applicable to isometric 2d games).

In such cases, it's common to use WebGL's depth buffer (which is supplied via `gl_Position.z` in the vertex shader), in combination with a "depth test" ([`gl.depthFunc(gl.LESS)`](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/depthFunc)). This essentially accomplishes the same thing as `sortableChildren`, but it sorts on a *per-pixel* basis. This technique hinges on the user's ability to provide depth values for each pixel of a sprite's texture (typically derived from world coordinates).

Is it possible to accomplish this in PixiJS?

Also, I have another question regarding a more advanced use-case... Using the technique described above can still have problems when layering transparent objects, as transparency is still order-dependent. The desired outcome is called "order independent transparency". People have solved this by writing a shader that builds a linked-list of overlapping pixels, using that (in combination with depth buffer values) to calculate the correct color in the fragment shader. My assumption is that PixiJS doesn't support such advanced functionality, but maybe there's an escape hatch for advanced plugins or something? Are you aware of anybody who has done this sort of thing before with PixiJS?

#### References for "order independent transparency":
- [Explanation of the algorithm](https://github.com/gangliao/Order-Independent-Transparency-GPU#linked-list-sorting)
- [Video describing how it works](https://www.youtube.com/watch?v=7EiCGEgb_No&t=4247s)
- [Example shader code implementing this strategy](https://github.com/nvpro-samples/vk_order_independent_transparency/blob/master/oitLinkedList.frag.glsl)

**Comment by ivanpopelyshev** [comment]
You need at least WebGL2 for this. I didnt see order independent transparency for pixi yet, but i managed to use usual z-buffer with transparency on old devices to overcome low fillrate problem. 

Also, for isometry its possible to sort elements and cut sprites into several if needed, you need good scanline algorithm. Unfortunately i lost the example.

Everything discussed is S-tier plugin, you have to pay for it ))

**Comment by shirecoding** [comment]
did you manage to get depth test for pixi v8 with transparency blending or alpha test?
